README
MailChimp 3.0 version 0.01
======================

This is the initial implentation of the MailChimp API version 3.0 in Perl.

It's pretty basic, but allows you to create a campaign, fill it with content,
and sent it to a list. It expects templates and lists to have already been
created on MailChimp.

Instead of using somethinbg big and heavy to implement, like Plack or Moose,
this module is built using Perl Curl. 

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  local::lib; # sets up a local lib at ~/perl5
  WWW::Curl::Easy;
  JSON;
  Digest::MD5;

local::lib allows implementation on Reclaim and probably not otherwise needed,
and if it breaks your implementation you can comment it out.

Digest::MD5 is needed to add emails to the mailing lint, but since I haven't implemented
that, is isn't required either.


=head1 NAME

MailChimp - Interface for MailChimp API 3.0

See https://mailchimp.com/developer/marketing/api/

=head1 SYNOPSIS

use MailChimp;

my $account = MailChimp->new({
		username => 'username email',
		datacenter => 'us18',
		version => '3.0',
		url => 'api.mailchimp.com',
		apikey => 'apikey'
});

$account->campaigns();      # List campaigns, etc
$account->templates();
$account->lists();

my $listid = "375520a149";  # Can also be defined using API but I haven't built that in yet
my $template_id = 69461;

	my $campaign = MailChimp::Campaigns->new({
		account => $account,
		type => 'regular',
		recipients => {
		    list_id => $listid
	    },
		settings => {
		    subject_line => 'Test',
		    preview_text => 'Text',
		    title => '',
		    from_name => 'moi',
		    reply_to => $account->{username},
		    to_name => "",
		    folder_id => "",
		    auto_fb_post => [],
		    template_id => $template_id
		},
		content_type => 'template'
	});
  
	$campaign->create();    # I prefer to keep this as an extra step 

	print $campaign->to_string();

  $campaign->add_content({
		html => "Some test HTML content",
		plain_text => "Some test text content"
	});

	$campaign->send();

=head1 DESCRIPTION

Implements a basic campaign creation and send topredefined email list in MailChimp.
Created by Stephen Downes to support the gRSShopper project.

=head2 EXPORT

None by default.

=head1 SEE ALSO

https://grsshopper.downes.ca

=head1 AUTHOR

Stephen Downes <lt>stephen@downes.ca<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2021 by Stephen Downes

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.


=cut  




Changes
Revision history for Perl extension NewModule.

0.01  Mon Jan 25 11:55:50 2021
	- original version; created by Stephen Downes with options
		-AXc -n NewModule




lib/MailChimp/Lists.pm
package MailChimp::Lists;

use 5.010001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use NewModule ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';


sub new {
    my $class = shift;
    my ( $list ) = @_;

    my $self = bless {
        list => $list->{list},
        account => $list->{account},
        id => $list->{id},
    }, $class;    

    print "List ID: ",$self->{list}->{id};
    my $request = MailChimp::Request->new({
        account => $self->{account},
        request=>'lists/'.$self->{list}->{id}
    });
    my $response = $request->submit();
    while (my ($lx,$ly) = each %$response) {
         $self->{$lx} = $ly;
        }

    return $self;
}

sub to_string {

    my $self = shift;
    return "List:" . sprintf("%s %s",$self->{list}->{id},$self->{account}->{datacenter});

}

lib/MailChimp/Request.pm
package MailChimp::Request;

use 5.010001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use NewModule ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';


sub new {
    my $class = shift;
    my ( $request ) = @_;


    my $self = bless {
        account => $request->{account},
        method => $request->{method},
        request => $request->{request},
        body => $request->{body},
    }, $class;
print "Request: ".$self->{request},"<p>";
    return $self;
}



# method can be GET or POST
# url is API url
# body is JSON encoded payload for POST requests

sub submit {

    my $self = shift;
    my $method = $self->{method};       # GET, POST, PUT
    my $request = $self->{request};     # 'lists','campaigns', etc
    my $body = $self->{body};           # hash of request content
    my $account = $self->{account};     # user account
    my $apikey = $account->{apikey};
#	my ($method,$url,$body) = @_;


    unless ($request) { $request = "ping"; }        # Default request    
    unless ($method)   { $method = "GET"; }         # Default method
    

    my $endpoint = sprintf("https://%s.%s/%s/%s",$account->{datacenter},$account->{url},$account->{version},$request);

# print "Method: $method Request: $endpoint<p>";

    my $curl = WWW::Curl::Easy->new;

    #my $url = $endpoint;
	#/$listid/members/" . Digest::MD5::md5(lc($email));

    $curl->setopt(WWW::Curl::Easy::CURLOPT_HEADER,0);
    $curl->setopt(WWW::Curl::Easy::CURLOPT_URL, $endpoint);

    $curl->setopt(WWW::Curl::Easy::CURLOPT_VERBOSE, 1);
    $curl->setopt(WWW::Curl::Easy::CURLOPT_USERPWD, 'user:' . $apikey);
    #$curl->setopt(WWW::Curl::Easy::CURLOPT_HTTPHEADER, ['Content-Type: application/json']);
    $curl->setopt(WWW::Curl::Easy::CURLOPT_TIMEOUT, 20);
    $curl->setopt(WWW::Curl::Easy::CURLOPT_CUSTOMREQUEST, $method);
    $curl->setopt(WWW::Curl::Easy::CURLOPT_SSL_VERIFYPEER, 0);
    if ($body) { $curl->setopt(WWW::Curl::Easy::CURLOPT_POSTFIELDS, $body); }

    # A filehandle, reference to a scalar or reference to a typeglob can be used here.
    my $response_body;
    $curl->setopt(WWW::Curl::Easy::CURLOPT_WRITEDATA,\$response_body);

    # Starts the actual request
    my $retcode = $curl->perform;

    # Looking at the results...
    if ($retcode == 0) {

		# print("Transfer went ok<p>");
		my $response_code = $curl->getinfo(WWW::Curl::Easy::CURLINFO_HTTP_CODE);

		# print "Received response: $response_body\n";

		use JSON::Parse 'parse_json';
		my $json = $response_body;
		if ($json) {
			my $perl = parse_json ($json);
			return $perl;
		} else {
			return $response_code;
		}
    } else {
        # Error code, type of error, error message
        my $message = "An error happened: $retcode ".$curl->strerror($retcode)." ".$curl->errbuf."\n";
		return $message;
	}

}



lib/MailChimp/Campaigns.pm
package MailChimp::Campaigns;

use 5.010001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use NewModule ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

sub new {
    my $class = shift;
    my ( $campaign ) = @_;

    my $self = bless {
        campaign => $campaign->{campaign},
        list => $campaign->{list},
        account => $campaign->{account},
        settings => $campaign->{settings},
        recipients => $campaign->{recipients},
        type => 'regular',
        id => $campaign->{id},
        'content-type' => 'template',
    }, $class;    

    return $self;
}


# Get campaign
# We are referring to a specific MailChimp campaign, get it from MailChimp

sub get {

    my $self = shift;

    if ($self->{id}) {         
        print "Campaign ID: ",$self->{id};
        my $request = MailChimp::Request->new({
            account => $self->{account},
            request=>'campaigns/'.$self->{id}
        });
        my $response = $request->submit();
        while (my ($lx,$ly) = each %$response) {
            $self->{$lx} = $ly;
            }
        
        return $self;
    }
}

sub create {

    my $self = shift;

    # Campaign needs an account and a list before it can be created
    die "Cannot create a campaign without an account" unless ($self->{account});
    die "Cannot create a campaign without specifying a list" unless ($self->{recipients}->{list_id});
 #   die "Cannot create a campaign without specifying a template" unless ($self->{settings}->{template_id});
    die "Cannot create a campaign without an email" unless
        (($self->{settings}->{reply_to}) || ($self->{account}->{email}));

    my $email = $self->{settings}->{reply_to} || $self->{account}->{email};
    my $new_campaign = {
		type => 'regular',
		recipients => {
		    list_id => $self->{recipients}->{list_id}
		},
		settings => {
			subject_line => 'Test',
			preview_text => 'Text',
			title => '',
			from_name => 'moi',
			reply_to => $email,
			to_name => "",
			folder_id => "",
			auto_fb_post => [],
			template_id => $self->{settings}->{template_id}
		},
		content_type => 'template'
	};

	use JSON;
	my $json_text = encode_json $new_campaign;

    my $request = MailChimp::Request->new({
        account => $self->{account},
        request=>'campaigns/',
        method => 'POST',
        body => $json_text
    });
    my $response = $request->submit();

    # Store values in campaign object
    while (my ($lx,$ly) = each %$response) {
        $self->{$lx} = $ly;
    }
    return $self;
}

sub add_content {

    my $self = shift;
    my $data = shift;
 

    die "Campaign must be created (use campaign->create();) on mailchimp before content can be added"
        unless ($self->{id});
    die "Content not defined in add_content()" unless ($data);
    
	# my $id = "fbb57afe8e";	# campaign id
	# my $id = $self->{id};	    # campaign id


	use JSON;
	my $json = encode_json $data;


    my $request = MailChimp::Request->new({
        account => $self->{account},
        request=> sprintf("campaigns/%s/content",$self->{id}),
        method => 'PUT',
        body => $json
    });
    my $response = $request->submit();

}

sub send {

    my $self = shift;
	print "Sending campaign<p>";

    my $request = MailChimp::Request->new({
        account => $self->{account},
        request=> sprintf("campaigns/%s/actions/send",$self->{id}),
        method => 'POST',
    });
    my $response = $request->submit();

		use Data::Dumper;
		print "<pre>";
		print Dumper($response);
			print "</pre>";		
		exit;

}

sub to_string {

    my $self = shift;
    return "Campaign:" . sprintf("%s : %s (%s)",$self->{create_time},$self->{id},$self->{account}->{apikey});

}

1;

lib/MailChimp/Templates.pm
package MailChimp::Templates;

use 5.010001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use NewModule ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);


our $VERSION = '0.01';



sub new {
    my $class = shift;
    my ( $template ) = @_;

    my $self = bless {
        template => $template->{template},
        account => $template->{account},
        id => $template->{id},
    }, $class;    

    return $self;
}

# Get template
# We are referring to a specific MailChimp template, get it from MailChimp

sub get {

    my $self = shift;

    if ($self->{id}) {         
        print "Template ID: ",$self->{id};
        my $request = MailChimp::Request->new({
            account => $self->{account},
            request=>'templates/'.$self->{id}
        });
        my $response = $request->submit();
        while (my ($lx,$ly) = each %$response) {
            $self->{$lx} = $ly;
            }
        
        return $self;
    }
}


sub to_string {

    my $self = shift;
    return "List:" . sprintf("%s %s",$self->{list}->{id},$self->{account}->{datacenter});

}

lib/MailChimp.pm
package MailChimp;

use 5.010001;
use strict;
use warnings;

use local::lib; # sets up a local lib at ~/perl5
use warnings;

use WWW::Curl::Easy;
use JSON;
use Digest::MD5;

use MailChimp::Request;
use MailChimp::Lists;
use MailChimp::Campaigns;
use MailChimp::Templates;
require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use NewModule ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';



sub new {
    my $class = shift;
    my ( $account ) = @_;

    my $self = bless {
        datacenter => $account->{datacenter},
        url => $account->{url},
        version => $account->{version},
        apikey => $account->{apikey},
    }, $class;

    return $self;
}

sub test {

  my $self = shift;
  my $request = MailChimp::Request->new({account => $self});
  return $request->submit();


}


sub campaigns {

  my $self = shift;
  my $count = shift;
  my $request = 'campaigns';
  if ($count) { $request .= "?count=$count"; }
  my $request = MailChimp::Request->new({account => $self,request=>$request});
  return $request->submit();

}

sub lists {

  my $self = shift;
  my $count = shift;
  my $request = 'lists';
  if ($count) { $request .= "?count=$count"; }
  my $request = MailChimp::Request->new({account => $self,request=>$request});
  return $request->submit();

}

sub templates {

  my $self = shift;
  my $count = shift;
  my $request = 'templates';
  if ($count) { $request .=  "?count=$count"; }
  my $request = MailChimp::Request->new({account => $self,request=>$request});
  return $request->submit();

}


1;
__END__


=head1 NAME

MailChimp - Interface for MailChimp API 3.0

See https://mailchimp.com/developer/marketing/api/

=head1 SYNOPSIS

use MailChimp;

my $account = MailChimp->new({
		username => 'username email',
		datacenter => 'us18',
		version => '3.0',
		url => 'api.mailchimp.com',
		apikey => 'apikey'
});

$account->campaigns();      # List campaigns, etc
$account->templates();
$account->lists();

my $listid = "375520a149";  # Can also be defined using API but I haven't built that in yet
my $template_id = 69461;

	my $campaign = MailChimp::Campaigns->new({
		account => $account,
		type => 'regular',
		recipients => {
		    list_id => $listid
	    },
		settings => {
		    subject_line => 'Test',
		    preview_text => 'Text',
		    title => '',
		    from_name => 'moi',
		    reply_to => $account->{username},
		    to_name => "",
		    folder_id => "",
		    auto_fb_post => [],
		    template_id => $template_id
		},
		content_type => 'template'
	});
  
	$campaign->create();    # I prefer to keep this as an extra step 

	print $campaign->to_string();

  $campaign->add_content({
		html => "Some test HTML content",
		plain_text => "Some test text content"
	});

	$campaign->send();

=head1 DESCRIPTION

Implements a basic campaign creation and send topredefined email list in MailChimp.
Created by Stephen Downes to support the gRSShopper project.

=head2 EXPORT

None by default.

=head1 SEE ALSO

https://grsshopper.downes.ca

=head1 AUTHOR

Stephen Downes <lt>stephen@downes.ca<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2021 by Stephen Downes

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.


=cut
